clients/unshaded/src/main/java/tachyon/client/block/TachyonBlockStore.java

```diff
@@ -90,17 +91,31 @@ public BufferedBlockInStream getInStream(long blockId) throws IOException {
       // TODO(calvin): Fix this RPC.
       BlockInfo blockInfo = masterClient.getBlockInfo(blockId);
       // TODO(calvin): Get location via a policy.
-      if (blockInfo.locations.isEmpty()) {
+      String localHostName = NetworkAddressUtils.getLocalHostName(ClientContext.getConf());
+      boolean hasLocalWorker = mContext.hasLocalWorker();
+      NetAddress workerNetAddress = null;
+      for (BlockLocation location : blockInfo.locations) {
+        workerNetAddress = location.getWorkerAddress();
+        boolean isLocalWorker = workerNetAddress.getHost().equals(localHostName);
+        if (hasLocalWorker) {
+          if (isLocalWorker) {
+            // Prefer local block if there exists a local worker.
+            break;
+          }
+        } else if (!isLocalWorker) {
+          // There is no local worker, and this is a remote block.
+          break;
+        }
+      }
+      if (workerNetAddress == null) {
         // TODO(calvin): Maybe this shouldn't be an exception.
-        throw new IOException("No block " + blockId + " is not available in Tachyon");
+        throw new IOException("Block " + blockId + " is not available in Tachyon");
       }
       // TODO(calvin): Investigate making this a Factory method
-      NetAddress workerNetAddress = blockInfo.locations.get(0).getWorkerAddress();
       InetSocketAddress workerAddr =
           new InetSocketAddress(workerNetAddress.getHost(), workerNetAddress.getDataPort());
-      if (NetworkAddressUtils.getLocalHostName(ClientContext.getConf()).equals(
-          workerAddr.getHostName())) {
-        if (mContext.hasLocalWorker()) {
+      if (localHostName.equals(workerAddr.getHostName())) {
+        if (hasLocalWorker) {
```

Comment:
Should some flag be set earlier (in the for loop), so that these conditions just examine that flag? I'm asking because in the loop, this information is already known.